### ISSUE TITLE
Windows installer stuck on 'Installing MeshCentral Service' step

### ISSUE THREAD

**Issue by theultramage** [Ylianst/MeshCentral#6331]
**Describe the bug**
I attempted to install MeshCentral for windows using the official installer. It deployed node.js and then proceeded to create the MeshCentral windows service. The installer just kinda stopped there, with no activity, no error message, no appcrash, no log files. Trying the 'Reinstall' option led to the same situation.

**To Reproduce**
1. Download latest MeshCentralInstaller.exe 
2. Proceed with the setup
3. See it stuck on 'Installing MeshCentral Service'
4. See event log stating that the new service was created successfully.

**Expected behavior**
The installer should progress through all of its steps to completion.

**Screenshots**
It's literally just the installer window stuck in that step doing nothing with all buttons greyed out.

**Server Software**
 - OS: Windows Server 2012r2
 - Virtualization: none
 - Network: LAN
 - Version: 1.1.27
 - Node: 18.19.0

**Additional context**
The same issue was reported on Jun 20 in https://github.com/Ylianst/MeshCentral/issues/6193#issuecomment-2180920060 but it got lost in the noise.

**Comment by si458** [comment]
its hasnt got lost, we just cant fully replicate the issue every single time

you are also using an old EOL server OS (2012r2)

the way to fix this issue is when it gets stuck at that step,
simply open up the windows services, then start the meshcentral service,
then randomly the step carries on

for some unknown reason, its just doesnt start the service after it install it, no idea why?

also the meshcentralinstaller is using an old version of `node-windows` which installs an old exe to help autostart nodejs,
so this might be part of the problem

**Comment by si458** [comment]
im just setting back up my server 2022 vm to have another look
(im still waiting for a donation to purchase a new SSD ❤️ https://www.si458.co.uk/2024/01/05/donation/)
i just discovered you run the meshcentralinstaller with a debug/log command and it will spit out info to help
so from the command line run `meshcentralinstaller.exe -log`, and this should create a `log.txt` file next to the installer
once it gets to that screen, give it say 2mins, then open the log file and check its out
can you also post the output too for me please?

**Comment by theultramage** [comment]
Mainstream support only ended some 10 months ago and it's only been security fixes for years, so I don't think that has anything to do with it. Extended support continues until 2026. If there's been an unintended OS compatibility break then it's by the installer. But more likely than not it's unrelated to the OS version.

Anyway, I used ProcMon during a Reinstall operation. I saw that `winservice.js --install` causes daemon.js to notice that winsw.exe is already copied and tries to record an eventlog warning, but that does not happen for some reason. Here's the commandline (that I confirm works from cmd when run manually): `C:\Windows\system32\cmd.exe /d /s /c ""C:\Program Files\Open Source\MeshCentral\node_modules\node-windows\bin\elevate\elevate.cmd" eventcreate /L APPLICATION /T WARNING /SO "MeshCentral Monitor" /D "The process cannot be installed again because it already exists." /ID 1000"`. Then winservice.js reacts to 'alreadyinstalled' by logging `StdOut: MeshCentral service already installed.` and doing process.exit(). That is the last observable activity.

If I manually delete the service and delete the 'daemon' subdirectory, the reinstallation completes fine.
If I uninstall everything and reinstall everything again, it also works fine. So unfortunately I am not able to reproduce the initial failure.

In summary, I observed the following:
 * node-windows's eventlog module fails to invoke the cmd subjob properly, causing daemon.js errors to not be logged to OS.
 * meshcentral.js on line 417 contains copy-pasted message handling from winservice.js which does process.exit() on error with no ui update, just logging, which may be the reason why the ui hangs.
 * one of my guesses why the service fails to start is, if the service creation happens asynchronously / with early exit, and due to server load the service is not yet ready when attempting to start it.

**Comment by si458** [comment]
this is the problem i was having, i couldnt replicate the issue every time!
the only time i could get it to cause this error was a completely fresh install of server os with all updates applied.
i had to remember to take a snapshot and revert back the snapshot every time after i attempted to install meshcentral
because every 2nd time of installing meshcentral, it would work no problem
but the issue i spotted was on a FRESH CLEAN OS where its never been installed before, not an existing OS running already

**Comment by si458** [comment]
ok so i cant get it to not install on my server 2022 vm BUT i got it to not install on my server 2012r2 vm!
attached is a screenshot from the logs showing it THINKS it installed the service and started the service
![image](https://github.com/user-attachments/assets/770133af-5473-4bae-a21c-3439d4db26ef)
but you can see from this next screenshot, the service did indeed install just never started,
or if it did start it stopped straight away for some reason?
![image](https://github.com/user-attachments/assets/071a6156-9a81-4365-8b0f-22e68bdaf081)

**Comment by theultramage** [comment]
Oh, that is good to know. Do you have any idea how node-windows creates and starts services? Does it use winapi, rpc, or invoke service.exe? I have a faint memory of realizing that service,exe can exit with success before the operation actually completes on the back-end, which would cause the followup 'start' command to run before the service is fully registered. It would be interesting to add a 5-second pause between creation and start, to see if that's the case.
If you want to see the nitty-gritty details, capture the moment using Sysinternals Process Monitor. Especially if all the actions are done by invoking cmd.exe, all the process creation events and their timing would be visible. It produces very noisy captures though, which require a bunch of filtering.

**Comment by si458** [comment]
im looking into it now as ive eaten my tea and before bed! but if on look tonight i will update tomorrow!

**Comment by theultramage** [comment]
There's a lot of variables. 2012 vs 2022 might be about the specific quirks of service processing and the related commandline utilities. The javascript modules used might be outdated, as pointed out earlier, so perhaps it involves incorrect usage of the apis. Would be even crazier if a recent change to a module said 'removed all compatibility tweaks for older OS versions'. Can't really tell at this point. Fortunately the workaround of re-running the install with the service files removed did help, so for me it's a low priority issue.

**Comment by si458** [comment]
after checking the code, its installing the service and once installed then asking it to start 
so it says its installed (which it is), and then says it started BUT IT DIDNT START!
what seems to be happening is its taking a long time to do the install step and i dont know why?
im watching it using procexp64 and it just seems to take forever to close the install,
so oversally when its running start afterwards, its not actually starting it because it hasnt finished installing it
so i think its a bug with `winsw` and its exe, but sadly `node-windows` is using a very old version of `winsw`

**Comment by theultramage** [comment]
I don't know if winsw uses the same api as the builtin sc / service.exe, but as an example, 'sc stop X' will return immediately and report the service as 'stop pending', instead of waiting and only returning when the service actually stops. This is why a batch script to restart a service by using 'sc stop & sc start' will malfunction. I assume service creation using sc also behaves this way. Opposite to this, the NET start / stop command waits to completion.

**Comment by si458** [comment]
from what i can see it just calls the `meshcentral.exe install` which is actually `winsw.exe` (and under the hood ITS install)
so its the winsw.exe that seems to be failing/taking too long and dont know why?
but it is weird why `node-windows` is finishing straight away tho saying its installed
even tho its actually still running in the background? still looking into it

**Comment by si458** [comment]
so.... just checked the source code for `node-windows` and discovered this...
```
// BEGIN SUPER AWFUL HACK TO GET AROUND WINSW.EXE ISSUE! REPLACE ASAP!!!!
// winsw.exe immediately responds with nothing, indicating success, even though
// it continues processing with the "install" method.
var sleep = function(period){
  var st = new Date().getTime();
  while(new Date().getTime() <= st+(period*1000)){}
  return;
};
```
so erm maybe we need to wait after its done the install for like 10 seconds before moving on and trying to start?
as it already does a 2 second wait to begin with when we do the install
```
me.execute('"'+path.resolve(dir,me._exe)+'" install',function(){
    sleep(2);
    me.emit('install');
});
```

**Comment by theultramage** [comment]
Ah. I saw that block of code but I skipped over it, thinking surely it can't be that.
In my faulty start/stop script, I added a loop that queried the service status and parsed out the status text, looking for 'started'. Though I don't know if that applies here, since there's no status code for a service that's being created. Either it uses one of the other ones, or the service simply does not exist until it's ready (so querying for existence, or for 'stopped' might work).

winsw is 1.17 (03/2015) but it does have a 'status' command which might be usable. I tried looking at winsw github to figure out if v3 does anything different, but their manual says nothing about this issue and from the code it is not obvious if its service creation waits to completion or not. I thought about asking them but the only listed means of communication is the issue tracker.

**Comment by si458** [comment]
from my testing, i took the setup back to basics,
installed nodejs,
`npm install meshcentral`
`npm install node-windows@0.1.14`
then tried to run the install function to create the windows service
`node node_modules/meshcentral/winservice.js --install`
and madly enough, if you watch the output, it displays the `im installed` and `ive started` logs
BUT
checked procexp64 and indeed the first `meshcentral.exe install` was still running in the background,
and from my timing it takes about 60seconds to close, and didnt actually start the service!

so it does seems to be a bug with winsw and not meshcentral

so what im thinking of doing is after we do the install of meshcentral from npm and install of node-windows from npm, the meshcentralinstaller and the JS code maybe just download directly from winsw from github the latest binary and replace the file in `node_modules/node-windows/bin/winsw/` (maybe set a FIXED version instead)

**Comment by si458** [comment]
ok just tried my comment above about downloading newer winsw and calling install, but that also doesnt work :(
still the same problem where it hangs on the install for like 60 seconds!

**Comment by si458** [comment]
ok i think ive fixed the issue but it means we have to wait for a new release of meshcentral for the issue to be fully resolved

whats happening is when it `node-windows` runs the winsw install `meshcentral.exe install` in our case,
it seems to be hanging for about 20(ish) seconds after its fully installed the service and ready to go!
no idea why and dont want to spend hours looking throught all the winsw source code and recompiling!

but whats also happening is because `winservice.js` is calling `install` and then `start` staight afterwards,
so its delaying the `winsw` by another 20(ish) seconds (so 20x2=40seconds)

`node-windows` also has a delay of 2 seconds to make sure that the `install` process completes before it actually returns the `install` event where `winservice.js` then starts the service!

so oversally its not starting properly because its not actually finished, and the meshcentralinstaller assumes it finished installing because it returned OK, even tho its not actually finished!

SO...

what ive done is
1. changed the `winservice.js` so install only installs and doesnt start
2. put a delay of 1 second after every return from `node-windows` on events
3. added a delay of 30 seconds after the meshcentralinstaller THINKS winsw actually installed correctly
4. then it starts meshcentral
5. then it checks its started and carrys on

so this is good news!
but steps 3 is installer code, where as step 1+2 is meshcentral code!
so the installer downloads the latest version from npm which would include the fixes for 1+2

so this is why it cant be fixed right now we have to wait for a new version
OR
i only change the delay in the meshcentralinstaller to say 60 seconds and then carry on
so i wouldnt do step 1+2+4

**Comment by si458** [comment]
ok all fixed now i think!
its actually only waiting 20secs in total not 20+20!
so i just put a delay in of 30 seconds then rerun the start command to make sure its defo started then carry on!
so give this installer a try when you have chance please! 
[MeshCentralInstaller.zip](https://github.com/user-attachments/files/16676209/MeshCentralInstaller.zip)

**Comment by theultramage** [comment]
That is rough... I would prefer to poll the service in a loop, or, study winsw / query their devs to understand what's going on, but I understand that that costs time and effort and this silly bugreport already consumed a lot. Either way, thanks a lot for looking into it.

**Comment by si458** [comment]
sadly winsw isnt really support anymore, the last update of version 2 was last year, same with their BETA version 3,
so we would have to completely find a new way to start nodejs on windows on startup

do try the installer in the zip file above tho as i said ive put a delay of 30 seconds in and that seems to work a treat!
i just have no idea why when you say install, depending on what OS is installed,
it finishes straight away even tho its not actually finished!

i even checked running winsw directly without node-windows, but it still takes 8 seconds to say its installed the service?

the other thing is the checker in the installer actually checks the serverstate.txt file thats inside meshcentral,
NOT the service itself!

also we could do with improving the meshcentralinstaller,
like for example,
1. copy the installer to inside program files and add a menu item, then you dont need to keep looking for the installer file or redownloading it!
2. stop/start the service with the installer as thats not possible at the moment.

PS: if you wanted any of future features/help check in the [meshcentralinstaller](https://github.com/Ylianst/MeshCentralInstaller) repo and open them there!
